File: subprojects/docs/src/samples/userguide/plugins/buildService/java/buildSrc/src/main/java/WebServer.java
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.services.BuildService;
import org.gradle.api.services.BuildServiceParameters;

import java.net.URI;
import java.net.URISyntaxException;

public abstract class WebServer implements BuildService<WebServer.Params>, AutoCloseable {
    interface Params extends BuildServiceParameters {
        Property<Integer> getPort();

        DirectoryProperty getResources();
    }

    public WebServer() throws URISyntaxException {
        int port = getParameters().getPort().get();
        URI uri = new URI(String.format("https://localhost:%d/", port));

        // start the server ...

        System.out.println(String.format("Server is running at %s", uri));
    }

    @Override
    public void close() {
        // stop the server ...
    }
}

File: subprojects/docs/src/samples/userguide/plugins/buildService/kotlin/buildSrc/src/main/kotlin/WebServer.kt
import org.gradle.api.services.BuildService
import org.gradle.api.services.BuildServiceParameters
import org.gradle.api.provider.Property
import org.gradle.api.file.DirectoryProperty
import java.net.URI

abstract class WebServer : BuildService<WebServer.Params>, AutoCloseable {
    interface Params : BuildServiceParameters {
        val port: Property<Int>
        val resources: DirectoryProperty
    }

    init {
        val port = parameters.port.get()
        val uri = URI("https://localhost:$port/")

        // start the server

        println("Server is running at $uri")
    }

    override fun close() {
        // stop the server
    }
}
